plugins/python/configwrap.cpp

```diff
@@ -0,0 +1,209 @@
+  // a string.
+  //
+  // Since the configuration is read-only, values are cached in the implicit
+  // dictionary such that they can continue to be inspected.
+  //
+  // Python arguments expected:
+  //  name: the property to retrieve
+  //  type: the type to cast to (optional) and one of: int, float (for C++
+  //        double), or str (for C++ std::string)
+  //        tuple as standin for std::vector
+  //  coll: boolean, set to True if this is a collection of <type>
+
+  PyObject *pyname = nullptr, *type = nullptr;
+  int coll = 0;
+  if (PyTuple_Check(args)) {
+    if (!PyArg_ParseTuple(args, "U|Op:__getitem__", &pyname, &type, &coll))
+      return nullptr;
+  } else
+    pyname = args;
+
+  // cached lookup
+#if PY_VERSION_HEX >= 0x030d0000
+  PyObject* pyvalue = nullptr;
+  PyObject_GetOptionalAttr((PyObject*)pycmap, pyname, &pyvalue);
+#else
+  PyObject* pyvalue = PyObject_GetAttr((PyObject*)pycmap, pyname);
+  if (!pyvalue)
+    PyErr_Clear();
+#endif
+  if (pyvalue)
+    return pyvalue;
+
+  std::string cname = PyUnicode_AsUTF8(pyname);
+
+  // typed conversion if provided
+  if (type == (PyObject*)&PyUnicode_Type) {
+    if (!coll) {
+      try {
+        auto const& cvalue = pycmap->ph_config->get<std::string>(cname);
+        pyvalue = PyUnicode_FromString(cvalue.c_str());
+      } catch (...) {
+        PyErr_Format(PyExc_TypeError, "property \"%s\" is not a string", cname.c_str());
+      }
+    } else {
+      try {
+        auto const& cvalue = pycmap->ph_config->get<std::vector<std::string>>(cname);
+        pyvalue = PyTuple_New(cvalue.size());
+        for (Py_ssize_t i = 0; i < (Py_ssize_t)cvalue.size(); ++i) {
+          PyObject* item = PyUnicode_FromString(cvalue[i].c_str());
+          PyTuple_SetItem(pyvalue, i, item);
+        }
+      } catch (...) {
+        PyErr_Format(
+          PyExc_TypeError, "property \"%s\" is not a collection of strings", cname.c_str());
+      }
+    }
+  } else if (type == (PyObject*)&PyLong_Type) {
+    if (!coll) {
+      try {
+        auto const& cvalue = pycmap->ph_config->get<long>(cname);
+        pyvalue = PyLong_FromLong(cvalue);
+      } catch (...) {
+        PyErr_Format(PyExc_TypeError, "property \"%s\" is not an integer", cname.c_str());
+      }
+    } else {
+      try {
+        auto const& cvalue = pycmap->ph_config->get<std::vector<std::string>>(cname);
+        pyvalue = PyTuple_New(cvalue.size());
+        for (Py_ssize_t i = 0; i < (Py_ssize_t)cvalue.size(); ++i) {
+          PyObject* item = PyLong_FromString(cvalue[i].c_str(), nullptr, 10);
+          PyTuple_SetItem(pyvalue, i, item);
+        }
+      } catch (...) {
+        PyErr_Format(
+          PyExc_TypeError, "property \"%s\" is not a collection of integers", cname.c_str());
+      }
+    }
+  } else if (type) {
+    PyErr_SetString(PyExc_TypeError, "requested type not supported");
+  }
+
+  if (type)
+    return pyvalue; // may be nullptr
+
+  // untyped (guess) conversion
```

Comment:
Can't we introspect the value (using [this](https://www.boost.org/doc/libs/latest/libs/json/doc/html/ref/value/kind.html), or the `is_*` members) and just generate the correct type?